### PR TITLE
Add retention flag, call to update retention

### PR DESCRIPTION
**Pull request by laniehei**
Add ability to update retention on a namespace

### CODE REVIEW

File: app/namespace.go [hunk 1]
```diff
@@ -574,6 +574,28 @@ func NewNamespaceCommand(getNamespaceClientFn GetNamespaceClientFn) (CommandOut,
 						},
 					},
 				},
+				{
+					Name:    "retention",
```
Comment (feihuang) [inline app/namespace.go:578]: for consistency with other commands, we should do ocld namespace retention set --retention-days 30?

Comment (laniehei) [reply app/namespace.go:578]: done

File: app/flags.go [hunk 2]
```diff
@@ -10,6 +10,7 @@ import (
 const (
 	ServerFlagName          = "server"
 	ConfigDirFlagName       = "config-dir"
+	RetentionFlagName       = "retention-days"
```
Comment (mastermanu) [inline app/flags.go:13]: [Nit] RetentionDaysFlagName

File: app/flags.go [hunk 3]
```diff
@@ -32,6 +33,13 @@ var (
 		Hidden:   true,
 		Required: false,
 	}
+	RetentionFlag = &cli.StringFlag{
+		Name:     RetentionFlagName,
+		Usage:    "The retention of the namespace",
+		Aliases:  []string{"ret"},
```
Comment (mastermanu) [inline app/flags.go:39]: [Nit] tctl uses the alias "rd". Not sure if we want to align on that?
Same with the usage. tctl refers to it as the "workflow execution retention". Might be good to have both tcld and tctl use same description / verbiage if possible?

File: app/namespace.go (hunk 1 above)
Comment (mastermanu) [reply app/namespace.go:578]: [Nit] given that we are specifying a flag on a subcommand called "retention", having "retention" in the flag name and the command seems redundant

File: app/namespace.go [hunk 4]
```diff
@@ -574,6 +574,33 @@ func NewNamespaceCommand(getNamespaceClientFn GetNamespaceClientFn) (CommandOut,
...
+					Usage:   "The length of time a closed workflow will be preserved before deletion",
```
Comment (mastermanu) [inline app/namespace.go:579]: [Nit] re-word usage a bit as this is a command, not a flag (e.g. "Manages the length of time a closed workflow...")

File: app/namespace.go [hunk 5]
```diff
@@ -574,6 +574,33 @@ func NewNamespaceCommand(getNamespaceClientFn GetNamespaceClientFn) (CommandOut,
...
+					Aliases: []string{"r"},
+					Subcommands: []*cli.Command{{
+						Name:    "set",
+						Aliases: []string{"s"},
+						Usage:   "Set the length of time a closed workflow will be preserved before deletion per namespace",
+						Flags: []cli.Flag{
```
Comment (mastermanu) [inline app/namespace.go:585]: shouldn't this take the ResourceVersion flag as well?

File: app/flags.go (hunk 3 above)
Comment (mastermanu) [reply app/flags.go:39]: Also, we should mention that the unit is days (even if the name of the flag makes it obvious, nice to also have the usage/help be explicit too) (e.g. "Workflow execution retention in days" or "The retention of the namespace in days")

File: app/namespace.go [hunk 6]
```diff
@@ -574,6 +574,33 @@ func NewNamespaceCommand(getNamespaceClientFn GetNamespaceClientFn) (CommandOut,
...
+							RetentionFlag,
+							RequestIDFlag,
+						},
+						Action: func(ctx *cli.Context) error {
+							retention := ctx.Int(RetentionFlagName)
+							n, err := c.getNamespace(ctx.String(NamespaceFlagName))
+							if err != nil {
+								return err
+							}
+							if int32(retention) == n.Spec.RetentionDays {
```
Comment (mastermanu) [inline app/namespace.go:596]: [Nit] if there is no uint flag, might want to check that a negative value is not supplied

**Review by mastermanu — CHANGES_REQUESTED** [review]
mostly minor comments

File: app/namespace.go [hunk 7]
```diff
@@ -574,6 +574,33 @@ func NewNamespaceCommand(getNamespaceClientFn GetNamespaceClientFn) (CommandOut,
...
+								return fmt.Errorf("retention for namespace is already set at %d days", ctx.Int(RetentionFlagName))
+							}
+							n.Spec.RetentionDays = int32(retention)
+							return c.updateNamespace(ctx, n)
+						},
+					}},
+				},
```
Comment (mastermanu) [inline app/namespace.go:603]: Please add corresponding unit tests. See what we do for certificate filters, certificates and/or search attributes as an example

File: app/namespace.go (hunk 1 above)
Comment (laniehei) [reply app/namespace.go:578]: I will keep this as "retention-days" to allow for usage of the same flag name

File: app/flags.go (hunk 3 above)
Comment (laniehei) [reply app/flags.go:39]: done

File: app/namespace.go (hunk 4 above)
Comment (laniehei) [reply app/namespace.go]: done

File: app/namespace.go [hunk 8]
```diff
@@ -574,6 +574,33 @@ func NewNamespaceCommand(getNamespaceClientFn GetNamespaceClientFn) (CommandOut,
 						},
 					},
 				},
+				{
+					Name:    "retention",
+					Usage:   "The length of time a closed workflow will be preserved before deletion",
+					Aliases: []string{"r"},
+					Subcommands: []*cli.Command{{
+						Name:    "set",
```
[the remark this answers is not in the document]
Comment (laniehei) [reply app/namespace.go:582]: done

File: app/namespace.go (hunk 5 above)
Comment (laniehei) [reply app/namespace.go:585]: done

File: app/namespace.go (hunk 6 above)
Comment (laniehei) [reply app/namespace.go:596]: uint was available thanks!

File: app/namespace.go (hunk 7 above)
Comment (laniehei) [reply app/namespace.go:629]: added tests

File: app/flags.go [hunk 9]
```diff
@@ -32,6 +33,13 @@ var (
 		Hidden:   true,
 		Required: false,
 	}
+	RetentionFlag = &cli.StringFlag{
```
[the remark this answers is not in the document]
Comment (laniehei) [reply app/flags.go]: done

File: app/namespace.go (hunk 6 above)
Comment (laniehei) [reply app/namespace.go]: I turned this into a uint flag, but it looks like it turns it into "0" if negative, so checking for negative.

File: app/namespace.go [hunk 10]
```diff
@@ -574,6 +574,59 @@ func NewNamespaceCommand(getNamespaceClientFn GetNamespaceClientFn) (CommandOut,
...
+								return c.updateNamespace(ctx, n)
+							},
+						},
+						{
+							Name:    "get",
+							Aliases: []string{"g"},
+							Usage:   "Retrieve the length of time (in days) a closed workflow will be preserved before deletion for a given namespace",
+							Flags: []cli.Flag{
+								NamespaceFlag,
+								RequestIDFlag,
```
Comment (anekkanti) [inline app/namespace.go:617]: Don't think we need the request id flag for get command.

**Review by anekkanti — APPROVED** [review]

**Pull request merged**